### PR TITLE
[JENKINS-50669] RingBufferLogHandler improvement

### PR DESCRIPTION
**Pull request by neaguandrei101**
See [JENKINS-50669](https://issues.jenkins-ci.org/browse/JENKINS-50669).

### Proposed changelog entries

* Entry 1: Removed the deprecated constructor in hudson.util.RingBufferLogHandler
* Entry 2: Removed the reflexive functions that did not work for the DEFAULT_RING_BUFFER_SIZE
* Entry 3: Fixed previous implementations that used the deprecated constructor
* Entry 4: Added unit tests for RingBufferLogHandler
* Entry 5: Added the required method from the ticket to hudson.WebAppMain

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] JIRA issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

File: core/src/main/java/hudson/WebAppMain.java [hunk 1]
```diff
@@ -80,8 +80,7 @@
  */
 public class WebAppMain implements ServletContextListener {
 
-    // use RingBufferLogHandler class name to configure for backward compatibility
-    private static final int DEFAULT_RING_BUFFER_SIZE = SystemProperties.getInteger(RingBufferLogHandler.class.getName() + ".defaultSize", 256);
+    private static final int DEFAULT_RING_BUFFER_SIZE = 256;
```
Comment (MRamonLeon) [inline core/src/main/java/hudson/WebAppMain.java]: Remove this. You can use the property in the `RingBufferLogHandler` which hold the value for every piece of code that needs it

File: core/src/main/java/hudson/WebAppMain.java [hunk 2]
```diff
@@ -80,8 +80,7 @@
  */
 public class WebAppMain implements ServletContextListener {
 
-    // use RingBufferLogHandler class name to configure for backward compatibility
-    private static final int DEFAULT_RING_BUFFER_SIZE = SystemProperties.getInteger(RingBufferLogHandler.class.getName() + ".defaultSize", 256);
```
Comment (MRamonLeon) [inline core/src/main/java/hudson/WebAppMain.java]: It's fine to remove it

File: core/src/main/java/hudson/WebAppMain.java [hunk 3]
```diff
@@ -90,6 +89,11 @@
             }
         }
     };
+
+    //written for the requirements of JENKINS-50669
+    public int getDefaultRingBufferSize() {
```
Comment (MRamonLeon) [inline core/src/main/java/hudson/WebAppMain.java]: this method is returning the current size and should be static for evergreen-plugin be able to call it. Please, remove this method and we can call directly the static field in the `RingBufferLogHandler`

File: core/src/main/java/hudson/logging/LogRecorder.java [hunk 4]
```diff
@@ -148,7 +148,7 @@ public AutoCompletionCandidates doAutoCompleteLoggerName(@QueryParameter String
     }
 
     @Restricted(NoExternalUse.class)
-    transient /*almost final*/ RingBufferLogHandler handler = new RingBufferLogHandler() {
+    transient /*almost final*/ RingBufferLogHandler handler = new RingBufferLogHandler(256) {
```
Comment (MRamonLeon) [inline core/src/main/java/hudson/logging/LogRecorder.java]: You can use the static (now public) field of `RingBufferLogHandler` instead of hard-coded 256

File: core/src/main/java/hudson/util/RingBufferLogHandler.java [hunk 5]
```diff
@@ -35,25 +35,19 @@
  */
 public class RingBufferLogHandler extends Handler {
 
-    private static final int DEFAULT_RING_BUFFER_SIZE = Integer.getInteger(RingBufferLogHandler.class.getName() + ".defaultSize", 256);
```
Comment (MRamonLeon) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java]: instead of removing it, make it public to be used by evergreen-plugin

File: core/src/main/java/hudson/util/RingBufferLogHandler.java [hunk 6]
```diff
@@ -35,25 +35,19 @@
...
     private int start = 0;
     private final LogRecord[] records;
     private volatile int size = 0;
 
-    /**
-     * This constructor is deprecated. It can't access system properties with {@link jenkins.util.SystemProperties}
-     * as it's not legal to use it on remoting agents.
-     */
-    @Deprecated
-    public RingBufferLogHandler() {
```
Comment (MRamonLeon) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java]: Even though it's deprecated, it's good to keep it just in case it's being used by any plugin.

File: core/src/main/java/hudson/util/RingBufferLogHandler.java [hunk 7]
```diff
@@ -35,25 +35,19 @@
...
-        this(DEFAULT_RING_BUFFER_SIZE);
-    }
-
+    //suggested size 256
     public RingBufferLogHandler(int ringSize) {
         records = new LogRecord[ringSize];
     }
 
+    public synchronized int getRecordsLength() {
```
Comment (MRamonLeon) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java]: You can remove this method, it's returning the current length and it was not required by the Jira ticket.

**Review by MRamonLeon — CHANGES_REQUESTED** [review]
Some changes suggested to address the requirements on the Jira ticket. We need to provide a `static` way to get the default size. We recommend:
`RingBufferLogHandler.DEFAULT_RING_BUFFER_SIZE`

**Comment by neaguandrei101** [comment]
I have added https://github.com/jenkinsci/jenkins/pull/4448/commits/cd8b2f34987eadaf3b06ab5a492d0ba5e2297ebe.
I made the field public and made the according changes.
This was a great idea, I never really worked with public fields but this is so much cleaner.
Now everything is static.

File: core/src/main/java/hudson/util/RingBufferLogHandler.java [hunk 8]
```diff
@@ -35,7 +35,7 @@
  */
 public class RingBufferLogHandler extends Handler {
 
-    private static final int DEFAULT_RING_BUFFER_SIZE = Integer.getInteger(RingBufferLogHandler.class.getName() + ".defaultSize", 256);
+    public static final int DEFAULT_RING_BUFFER_SIZE = 256;
```
Comment (MRamonLeon) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java]: Why did you change this `= Integer.getInteger(RingBufferLogHandler.class.getName() + ".defaultSize", 256);` to just ` = 256`? 
With the former way, we can change the buffer size with a System property, chich could be interesting.

Comment (neaguandrei101) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java]: ```RingBufferLogHandler.class.getName() + ".defaultSize"``` was flawed for example because the .defaultSize property does not exist and the function attributed by default 256. https://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html#getInteger(java.lang.String,%20int)
That function is an reflexive and unnecessary .
From the ticket it is said "Currently, as an intermediate shim, I'm using reflection. In addition to being dirty, this is likely to become even more an issue to support Java 9 when using such hacks is being made harder and harder." I agree with that statement.
Attributing an integer is better in my opinion.

Comment (neaguandrei101) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java]: If I keep that previous function I need to add a int defaultSize field that is itself 256, that serves no purpose.

Comment (MRamonLeon) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java]: I'm pretty sure @batmat was refferring to [these lines](https://github.com/jenkinsci/evergreen-plugin/commit/5f42dc4e3cca1981dc7c5636b4139731c0d16ec4#diff-4a65f41c4b96b07b21810d8bd8a51161R82-R83) when he said _being dirty, this is likely to become even more an issue to support Java 9_ . There is nothing wrong with using a system property to get the default value, it has nothing to do with reflection either. 

The reflective call is https://github.com/jenkinsci/evergreen-plugin/commit/5f42dc4e3cca1981dc7c5636b4139731c0d16ec4#diff-4a65f41c4b96b07b21810d8bd8a51161R84 and the issue with Java9 and higher is https://github.com/jenkinsci/evergreen-plugin/commit/5f42dc4e3cca1981dc7c5636b4139731c0d16ec4#diff-4a65f41c4b96b07b21810d8bd8a51161R83

Please consider keeping the former behavior, but `public`.

Comment (MRamonLeon) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java]:
```suggestion
    public static final int DEFAULT_RING_BUFFER_SIZE = Integer.getInteger(RingBufferLogHandler.class.getName() + ".defaultSize", 256);	
```

File: core/src/main/java/hudson/logging/LogRecorder.java [hunk 9]
```diff
@@ -148,7 +148,7 @@ public AutoCompletionCandidates doAutoCompleteLoggerName(@QueryParameter String
     }
 
     @Restricted(NoExternalUse.class)
-    transient /*almost final*/ RingBufferLogHandler handler = new RingBufferLogHandler() {
+    transient /*almost final*/ RingBufferLogHandler handler = new RingBufferLogHandler(RingBufferLogHandler.DEFAULT_RING_BUFFER_SIZE) {
```
Comment (MRamonLeon) [inline core/src/main/java/hudson/logging/LogRecorder.java]: You can keep the previous line, the default size of the non-arg constructor is already `RingBufferLogHandler.DEFAULT_RING_BUFFER_SIZE`

File: core/src/main/java/hudson/WebAppMain.java [hunk 10]
```diff
@@ -80,16 +80,19 @@
  */
 public class WebAppMain implements ServletContextListener {
 
-    // use RingBufferLogHandler class name to configure for backward compatibility
-    private static final int DEFAULT_RING_BUFFER_SIZE = SystemProperties.getInteger(RingBufferLogHandler.class.getName() + ".defaultSize", 256);
-
-    private final RingBufferLogHandler handler = new RingBufferLogHandler(DEFAULT_RING_BUFFER_SIZE) {
+    private final RingBufferLogHandler handler = new RingBufferLogHandler(RingBufferLogHandler.DEFAULT_RING_BUFFER_SIZE) {
```
Comment (MRamonLeon) [inline core/src/main/java/hudson/WebAppMain.java]: You can call the no-arg constructor because it uses `RingBufferLogHandler.DEFAULT_RING_BUFFER_SIZE`
```suggestion
    private final RingBufferLogHandler handler = new RingBufferLogHandler() {
```

**Comment by neaguandrei101** [comment]
I did everything @MRamonLeon suggested by the letter in this commit.

File: core/src/main/java/hudson/WebAppMain.java [hunk 11]
```diff
@@ -80,16 +80,19 @@
  */
 public class WebAppMain implements ServletContextListener {
 
-    // use RingBufferLogHandler class name to configure for backward compatibility
-    private static final int DEFAULT_RING_BUFFER_SIZE = SystemProperties.getInteger(RingBufferLogHandler.class.getName() + ".defaultSize", 256);
-
-    private final RingBufferLogHandler handler = new RingBufferLogHandler(DEFAULT_RING_BUFFER_SIZE) {
+    private final RingBufferLogHandler handler = new RingBufferLogHandler() {
```
Comment (alecharp) [inline core/src/main/java/hudson/WebAppMain.java]: The no-arg constructor is marked as deprecated. We could use 

```suggestion
    private final RingBufferLogHandler handler = new RingBufferLogHandler(WebAppMain.getDefaultRingBufferSize()) {
```

what do you think?

**Review by alecharp — CHANGES_REQUESTED** [review]
Current state of the PR is using a `@Deprecated` constructor.

File: core/src/main/java/hudson/WebAppMain.java (hunk 11 above)
Comment (neaguandrei101) [reply core/src/main/java/hudson/WebAppMain.java]: I did what was suggested. It was foolish of me to commit using a deprecated function. As were my previous mistakes.
The required functionality was achieved after all with changes in just 2 places.^^
I hope this commit does the trick.

**Comment by neaguandrei101** [comment]
I need to find out what went wrong. Sorry for spamming your feed. One more commit.

**Comment by neaguandrei101** [comment]
No idea why the tests failed this time.

**Comment by timja** [comment]
> No idea why the tests failed this time.

The test failure is fixed on master

**Review by varyvol — APPROVED** [review]

**Review by fcojfernandez — APPROVED** [review]

**Review by alecharp — APPROVED** [review]

**Comment by batmat** [comment]
@MRamonLeon do the changes done match your expectations? Care to have another look given you're the last one blocking this now, thanks :-)!

**Review by oleg-nenashev — COMMENTED** [review]
IMHO we need either a public static field or a public static getter method. But we do not need both. My preference (not a strong opinion) is to only have a getter method. Whatever approach is selected, it would need some Javadoc for API consumers.

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Formally requesting a change according to the comment above. It is almost ready to go, just a minor tweak is needed

**Comment by neaguandrei101** [comment]
I made the field DEFAULT_RING_BUFFER_SIZE private and I exposed access to it with static getters. I believe this design does not create confusion and is as pretty as the public field was. 
I fulfilled the requirements from the ticket and I added a java-doc comment.
